### PR TITLE
63 githubauth

### PR DESCRIPTION
**Pull request by Nkoli**

### CODE REVIEW

File: .env.example [hunk 1]
```diff
@@ -1,8 +1,10 @@
...
-DB_USER = 'YOUR_DB_USERNAME
-DB_PASSWORD = 'YOUR_DB_PASSWORD
+DB_NAME = 'Your postgres database name'
+DB_USER = 'Your postgres database user'
+DB_PASSWORD = 'Your postgres database password'
 CLOUDINARY_CLOUD_NAME = 'Your cloudinary account name'
 CLOUDINARY_API_KEY = 'Your cloudinary api key'
 CLOUDINARY_API_SECRET = 'Your cloudinary api secret'
+SOCIAL_AUTH_GITHUB_KEY = 'Your github app client id'
+SOCIAL_AUTH_GITHUB_SECRET = 'Your github app client secret'
```
Comment (Del-sama) [inline .env.example]: Add a new line at the end

File: app/models.py [hunk 2]
```diff
@@ -52,12 +52,18 @@ class User(models.Model):
...
+    )
 
     def __str__(self):
         return self.display_name
 
 
+class GithubAuth(models.Model):
+    github_user_id = models.CharField(max_length=200)
+    github_access_token = models.CharField(max_length=4096)
+    user_profile = models.OneToOneField(User, on_delete=models.CASCADE)
```
Comment (Del-sama) [inline app/models.py]: I think you can just call this `user`

Comment (Del-sama) [reply app/models.py]: You can also remove the github prefix from the other names, I think

File: app/oauth.py [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+def generate_github_access_token(github_client_id, github_client_secret, github_code):
+    """
+    create an access token to github oauth2
+    """
+    github_response = requests.post(
+        'https://github.com/login/oauth/access_token/',
+        data=json.dumps({
+            'client_id': github_client_id,
+            'client_secret': github_client_secret,
+            'code': github_code
```
Comment (Del-sama) [inline app/oauth.py:18]: What's this github code?

File: app/oauth.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+        }),
+        headers={'content-type': 'application/json'}
+    )
+
+    token = re.search(r'access_token=([a-zA-Z0-9]+)',
+                      github_response.content.decode('utf-8'))
+    if token is None:
```
Comment (Del-sama) [inline app/oauth.py]: You can also say `if not token`

File: app/oauth.py [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+        'https://github.com/login/oauth/access_token/',
+        data=json.dumps({
+            'client_id': github_client_id,
+            'client_secret': github_client_secret,
+            'code': github_code
+        }),
+        headers={'content-type': 'application/json'}
+    )
+
+    token = re.search(r'access_token=([a-zA-Z0-9]+)',
```
Comment (Del-sama) [inline app/oauth.py:24]: I'm not sure whether we need this regex, I'd liek to pair on this

File: app/oauth.py [hunk 6]
```diff
@@ -0,0 +1,54 @@
+import json
+import re
+
+import requests
+from oauth2_provider.models import AccessToken
+
+from .models import GithubAuth
+
+
+def generate_github_access_token(github_client_id, github_client_secret, github_code):
```
Comment (Del-sama) [inline app/oauth.py:9]: I think we should use better docstrings and we should also use python's `typing module` for type hinting
```
def generate_github_access_token(github_client_id: str, github_client_secret: str, github_code: str) -> str:
"""
Create a Github access token

:Params
     github_client_id: client id
     github_client_secret: client secret
     github_code: client code

:Return: parsed token
"""
```
Something like that

File: app/oauth.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    authenticate the user and return a django auth token
+    """
+    data = {
+        'grant_type': 'convert_token',
+        'client_id': client_id,
+        'client_secret': client_secret,
+        'backend': backend,
+        'token': token,
+    }
+    response = requests.post('http://127.0.0.1:8000/auth/convert-token/', data=data)
```
Comment (Del-sama) [inline app/oauth.py]: Is the url always going to be `localhost`?

**Review by Del-sama — COMMENTED** [review]
Let's look at this together on Sunday.
Side note: I like the way you arrange your imports.

File: app/models.py (hunk 2 above)
Comment (Nkoli) [reply app/models.py]: okay

File: app/oauth.py (hunk 3 above)
Comment (Nkoli) [reply app/oauth.py:18]: the code that's sent from the frontend. I presumed it'll need to be saved in a variable.

File: app/oauth.py (hunk 5 above)
Comment (Nkoli) [reply app/oauth.py:24]: okay then, what time?

File: app/oauth.py (hunk 6 above)
Comment (Nkoli) [reply app/oauth.py:9]: Okay, will do.

File: app/oauth.py (hunk 7 above)
Comment (Nkoli) [reply app/oauth.py]: thought about this but I really don't know how to make it work both in dev and prod.

**Pull request closed without merging**